### ISSUE TITLE
PDFs embedded via File block disappear once published

### ISSUE THREAD

**Issue by SeanDS** [WordPress/gutenberg#34738]
### Description

A few users of my internal WordPress instance have reported that the new [PDF inline embed feature](https://make.wordpress.org/core/2021/04/30/whats-new-in-gutenberg-10-5-28-april/) breaks when a post containing an embedded PDF document is published or subsequently edited.

Creating a new post, adding a `File` block then choosing a PDF document works in the editor - the PDF is shown in an embedded frame. Once the post is published, though, the front-end version does not display the embedded frame but only the Download button. And upon editing the post again, the editor displays an "invalid block" message for the `File`.

Interestingly, on my computer (Firefox 91 on Linux), I can create and edit posts containing embedded PDF documents without any issue - the embedded frame appears on the front end and later when editing. But if one of the users for whom it does not work edits my or any other post containing an embedded PDF, they break it when they save the post again.

A diff between revisions of such a post shows what breaks:

![image](https://user-images.githubusercontent.com/5225190/132702096-652497f8-ea31-428f-92a7-9ecdc4f3d02f.png)

(working version post-publication on left, new broken version on right)

It seems their editors are removing the `<object>` element before saving, whereas mine is not.

### Step-by-step reproduction instructions

On an affected platform (seems to be Firefox on Windows 10, at least), do the following:

1. Create a new post, adding an embedded PDF. Note the PDF is displayed inline.
2. Save the post. Note the published post on the front-end does not display the PDF any more, just the download button.
3. Edit the post. Note the invalid block error message.

### Environment info

WordPress 5.8.1, Gutenberg editor (whatever one is bundled with 5.8.1). Active theme is custom, found [here](https://github.com/Academic-Labbook/themes/tree/master/labbook).

Browsers: this does NOT seem to affect me, on Firefox 91.0.2 on Linux. Of two affected users that I've obtained the information from, they both use Windows 10, and Firefox (one reported 91.0.2, the other didn't specify a version).

Note: to avoid having to set up local accounts on my test server, I had affected users test this on a test server with all plugins disabled *except* a single sign-on plugin called wpDirAuth, which I expect has nothing to do with this problem. They still encountered the issue.

### Please confirm that you have searched existing issues in the repo.

Yes

### Please confirm that you have tested with all plugins deactivated except Gutenberg.

Yes

**Comment by SeanDS** [comment]
Here's the console error when a post with this issue is edited:

```
Block validation: Block validation failed for `core/file` ( 
Object { name: "core/file", icon: {…}, keywords: (3) […], attributes: {…}, providesContext: {}, usesContext: [], supports: {…}, styles: [], save: save(), apiVersion: 2, … }
 ).

Content generated by `save` function:

<div class="wp-block-file"><object class="wp-block-file__embed" data="https://logbooks.aei.uni-hannover.de/laserlogbook/wp-content/uploads/sites/17/2021/08/iss_rpn-009.pdf" type="application/pdf" style="width:100%;height:600px" aria-label="Embed of iss_rpn-009."></object><a href="https://logbooks.aei.uni-hannover.de/laserlogbook/wp-content/uploads/sites/17/2021/08/iss_rpn-009.pdf">iss_rpn-009</a><a href="https://logbooks.aei.uni-hannover.de/laserlogbook/wp-content/uploads/sites/17/2021/08/iss_rpn-009.pdf" class="wp-block-file__button" download>Download</a></div>

Content retrieved from post body:

<div class="wp-block-file"><a href="https://logbooks.aei.uni-hannover.de/laserlogbook/wp-content/uploads/sites/17/2021/08/iss_rpn-009.pdf">iss_rpn-009</a><a href="https://logbooks.aei.uni-hannover.de/laserlogbook/wp-content/uploads/sites/17/2021/08/iss_rpn-009.pdf" class="wp-block-file__button" download>Download</a></div> blocks.min.js:2:40465
```

And the abbreviated `core/file` object above:

```json
{
  "name": "core/file",
  "icon": {
    "src": {
      "key": null,
      "ref": null,
      "props": {
        "viewBox": "0 0 24 24",
        "xmlns": "http://www.w3.org/2000/svg",
        "children": {
          "key": null,
          "ref": null,
          "props": {
            "d": "M19 6.2h-5.9l-.6-1.1c-.3-.7-1-1.1-1.8-1.1H5c-1.1 0-2 .9-2 2v11.8c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V8.2c0-1.1-.9-2-2-2zm.5 11.6c0 .3-.2.5-.5.5H5c-.3 0-.5-.2-.5-.5V6c0-.3.2-.5.5-.5h5.8c.2 0 .4.1.4.3l1 2H19c.3 0 .5.2.5.5v9.5z"
          },
          "_owner": null
        }
      },
      "_owner": null
    }
  },
  "keywords": [
    "document",
    "pdf",
    "download"
  ],
  "attributes": {
    "id": {
      "type": "number"
    },
    "href": {
      "type": "string"
    },
    "fileName": {
      "type": "string",
      "source": "html",
      "selector": "a:not([download])"
    },
    "textLinkHref": {
      "type": "string",
      "source": "attribute",
      "selector": "a:not([download])",
      "attribute": "href"
    },
    "textLinkTarget": {
      "type": "string",
      "source": "attribute",
      "selector": "a:not([download])",
      "attribute": "target"
    },
    "showDownloadButton": {
      "type": "boolean",
      "default": true
    },
    "downloadButtonText": {
      "type": "string",
      "source": "html",
      "selector": "a[download]"
    },
    "displayPreview": {
      "type": "boolean"
    },
    "previewHeight": {
      "type": "number",
      "default": 600
    },
    "align": {
      "type": "string",
      "enum": [
        "left",
        "center",
        "right",
        "wide",
        "full",
        ""
      ]
    },
    "className": {
      "type": "string"
    },
    "anchor": {
      "type": "string",
      "source": "attribute",
      "attribute": "id",
      "selector": "*"
    }
  },
  "providesContext": {},
  "usesContext": [],
  "supports": {
    "anchor": true,
    "align": true
  },
  "styles": [],
  "apiVersion": 2,
  "title": "File",
  "description": "Add a link to a downloadable file.",
  "category": "media",
  "variations": [],
  "transforms": {
    "from": [
      {
        "type": "files",
        "priority": 15
      },
      {
        "type": "block",
        "blocks": [
          "core/audio"
        ]
      },
      {
        "type": "block",
        "blocks": [
          "core/video"
        ]
      },
      {
        "type": "block",
        "blocks": [
          "core/image"
        ]
      }
    ],
    "to": [
      {
        "type": "block",
        "blocks": [
          "core/audio"
        ]
      },
      {
        "type": "block",
        "blocks": [
          "core/video"
        ]
      },
      {
        "type": "block",
        "blocks": [
          "core/image"
        ]
      }
    ]
  }
}
```

**Comment by vtad** [comment]
This error occurs only for not admin users (role author, etc.)

**Comment by SeanDS** [comment]
> This error occurs only for not admin users (role author, etc.)

Ah yes, I just checked and I can confirm that this problem appears for the editor role but not administrator.

**Comment by getsource** [comment]
Thanks much for the report!
I was able to reproduce this as an author.

@pento Do you have any opinions on the best route here? I'm familiar with a little of the history of kses and `<object>` but not enough to know the current state / workarounds.

**Comment by pento** [comment]
Thanks for the ping, @getsource. That's a fun bug, I'm not sure how it got through: I thought the block fixture testing was supposed to catch things like this? @gziolo?

I have two suggestions for fixing this, neither of which are ideal, but they're both workable:

- Modify `render_block_core_file()` to check if the `<object>` tag is there when it's supposed to be. If it isn't, add it back in.
- Modify KSES to allow the `<object>` tag when the `type` attribute is set to `application/pdf`. KSES currently doesn't support this behaviour, but it would be a relatively straightforward modification of `wp_kses_split2()`.

The former has the advantages of being relatively simpler (it can be fixed entirely within Gutenberg), and it also retroactively fixes the issue for any sites that have already run into it. The latter is probably the more correct fix, as it ensures the block behaves the same for everyone, but nobody likes touching KSES. 🙂

I _wouldn't_ recommend allowing `<object>` tags without limitations: while most of the original problematic uses of it are no longer supported in browsers (Java applets, ActiveX, Flash, etc), it would be challenging to prove that there are _no_ potential security issues with allowing it for all object types.

**Comment by gziolo** [comment]
> That's a fun bug, I'm not sure how it got through: I thought the block fixture testing was supposed to catch things like this? @gziolo?

Block fixtures testing in Gutenberg operates only on JavaScript codebase so as far as I know, KSES is completely ignored.

**Comment by pento** [comment]
I've written a fix for this in [WP54261](https://core.trac.wordpress.org/ticket/54261), though it does need thorough reviewing. #35611 implements some preventative testing that will help avoid issues like this one in the future.

**Comment by SeanDS** [comment]
Thanks for looking into this. It seems like the fix is going to be non-negligible work, so can the developers consider rolling back this broken feature in the next release until it's fixed? The feature was dead on arrival. There must be thousands of regular users e.g. on WordPress.com running into this problem...

**Comment by pento** [comment]
Thanks for checking in, @SeanDS! I forgot to update this issue, here's the current status:

- The fix has landed in WordPress Core, it'll be included in the WordPress 5.9 release. WordPress 5.9 beta 1 is due out next week, if you're comfortable running a beta release on your site.
- Separately, the fix is currently being tested on WordPress.com, Automattic/wp-calypso#52728 will be updated when it lands there.

I agree that this process has been far from ideal, #35611 has been merged to help prevent situations like this occurring in the future. With WordPress 5.9 being scheduled for release soon, I don't think it's necessary to roll the feature back, as it will work properly as soon as sites are upgraded.

**Comment by SeanDS** [comment]
Ah excellent, I agree no rollback needed! Thanks for the (comparatively - given WP release cycles) quick fix - much appreciated, also for the original work on PDF embeds!
